## 1. C++调用普通的python函数

+ C++环境配置
  
```
INCLUDEPATH += D:\Python\Python39\include \
LIBS += D:\Python\Python39\libs\python39.lib \

记得把python3.dll和python3.9.dll放到exe同级目录中
```

+ C++代码

```cpp

#undef slots
#include <Python.h>
#include <tupleobject.h>
#define slots Q_SLOTS

int main(int argc, char *argv[])
{
    Py_Initialize();  //必要，初始化python解释器

    PyRun_SimpleString("import sys");
    PyRun_SimpleString(QString("sys.path.append('%1')")
                       .arg("D:\\qtproject\\call_python")
                       .toStdString().c_str());

    // 注册py文件
    PyObject * pModule = PyImport_ImportModule("py_func"); 
    if (pModule == nullptr) {
        qInfo() << "Error: load failed. "
                << QStringLiteral("py不存在或者无法正常运行，可能是python的环境或依赖库异常");
    }

    // 注册函数
    QString func_name = "show_cv";
    PyObject * pFunc = PyObject_GetAttrString(pModule, func_name.toStdString().c_str());
    if (pFunc == nullptr) {
        qInfo() << "Error: load failed. " << ", Func: "<< func_name;
    }

    PyObject *pArgs = PyTuple_New(2);
    PyTuple_SetItem(pArgs,0, Py_BuildValue("i", 45));
    PyTuple_SetItem(pArgs,1, Py_BuildValue("s", "3"));

    // 调用函数
    char * res_str = 0;
    PyObject *ret2 = PyObject_CallObject(pFunc, pArgs);
    PyArg_Parse(ret2, "s", &res_str);

    qInfo() << "result: " << QString::fromLatin1(res_str);

    Py_Finalize();

    return 0;
}
```

+ python代码

```py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys

print ("py: ", sys.version)
def show_cv(img, type):
    return "json_string" + type
```

## 2. 传递cv::Mat给Python程序

&emsp;&emsp;使用的是pyhon的numpy模块进行数据传输。

```cpp
/* img_data函数中的uchar *也可能是unsigned short等
        这要根据cv::Mat的图像类型来。比如深度图一般是16位
*/
uchar *call_python::img_data(const cv::Mat &img)
{
    int row = img.rows;
    int col = img.cols;
    int channel = img.channels();
    uchar *imgData = new uchar[row*col*channel];

    int irow = row,icol = col*channel;
    //判断图像的Mat数据是不是一行保存的
    if (!img.isContinuous())
    {
        icol *= irow;
        irow = 1;
    }
    int id = 0;
    for(int i = 0; i<irow; i++)
    {
        for(int j = 0; j<icol; j++)
        {
            imgData[id] = img.at<uchar>(i,j);
            id++;
        }
    }
    return imgData;
}

int main(int argc, char *argv[])
{
    uchar * p_img_data = cv_data(img);
    npy_intp Dims[3] = { img.rows, img.cols, img.channels()};
    PyObject *PyImg = PyArray_SimpleNewFromData(3, Dims, NPY_UBYTE, p_img_data);
    // 将numpy数组构造为tuple
    PyObject *pArgs = PyTuple_New(2);
    PyTuple_SetItem(pArgs,0, PyImg);
    PyTuple_SetItem(pArgs,1, Py_BuildValue("s", "3"));

    return 0;
}
```

```py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np
import sys
import json
import cv2

def show_cv(img, type):
    print (type)
    cv2.imshow("img", img)
    cv2.waitKey(3000)
    return "json_string"
```

#### 2.1 深度图

&emsp;&emsp;深度图中使用的是16位的矩阵（cv::Mat）。

```cpp
cv::Mat depth = cv::imread("D:\\remote\\img\\31_depth.png", CV_16UC1);

// 构造传递给python的参数
quint16 *img_data_1 = img_data_depth(img_depth);
npy_intp Dims_1[3] = { img_depth.rows, img_depth.cols, img_depth.channels()};
PyObject *PyImg_1 = PyArray_SimpleNewFromData(3, Dims_1, NPY_USHORT, img_data_1);
```

```py
# 读取深度图
depth_frame = cv2.imread("D:/remote/img/31_depth.png", cv2.CV_16UC1)
# cv裁切图片
cropped = depth_frame[y1:y2, x1:x2]

# cv -> numpy     16位互转
npy = np.asarray(depth_frame)
```

## 3. 项目开发过程中发现的问题

#### 问题1：

  python代码迭代中发现
  在命令行中执行py脚本（main函数中读取C++的深度图，然后调用被C++调用的接口），程序是ok的，可以正常返回结果。但C++中调用的话，PyObject_CallObject返回nullptr，无法获取执行结果。

  在```PyObject_CallObject```后加入```if (PyErr_Occurred()) { PyErr_Print();}```，打印如下堆栈信息

```
File "<__array_function__ internals>", line 200, in hstack
  File "D:\Python\Python39\lib\site-packages\numpy\core\shape_base.py", line 370, in hstack
    return _nx.concatenate(arrs, 1, dtype=dtype, casting=casting)
  File "<__array_function__ internals>", line 200, in concatenate
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 3 dimension(s) and the array at index 1 has 2 dimension(s)
```

  后来发现```np.hstack((a,b))```、```np.c_[a,b]```、```np.column_stack((a,b))```等都无法使用，会爆同样的错误。

#### 问题2：

  python代码迭代中发现
  这个问题同样操蛋
  在命令行中执行py脚本，程序是ok的，可以正常返回结果。但C++中调用后发现，返回的部分结果有误。

  一开始猜测和问题1同样，有些接口返回的值是有问题的。
  最后发现，在接口中将数据保存到本地后，再次读取，程序可以正常运行。代码如下：

```py
cv2.imwrite("D:\\depth.png", depth_img)
logger.debug(type(depth_img))  # <class 'numpy.ndarray'>
depth_img = cv2.imread("D:\\depth.png", cv2.CV_16UC1)
logger.debug(type(depth_img))  # <class 'numpy.ndarray'>
```

  尝试在代码中加入numpy矩阵的判等操作```(depth_img == depth_img_2).all()```，C++中获取如下错误```TypeError: function takes at least one argument```。对python不熟，暂时不继续跟踪了。

## 4. C++中获取python的log输出

#### 重定向到sys.stdout

```py
log_file = open("py.log", "w")
sys.stdout = log_file
```

+ 效果

脚本运行可以正常输出日志到文件，混合编程中无法输出log到文件中。被C++"劫持"了？

#### 使用 loguru

+ 安装

```py
pip install loguru      # 注意，安装的是loguru，而不是logger
```

+ 使用
  
```py
from loguru import logger
logger.add('./log-py.log', encoding="utf-8")
logger.debug("Happy logging with Loguru!")
```

+ 效果

日志文件log-py.log和qtcreator的应用程序输出界面中都可以看到python程序输出的log